### DISCUSSION TITLE
Mobile version of TShock?(1.3.0.7)

### DISCUSSION THREAD

**Discussion by lysenika** [Pryaxis/TShock#1721] [General]
Devs, can you optimize TShock for mobile platforms? Now its working, but with many-many lags and bugs.

**Comment by hakusaro** [comment]
You can make a feature request issue to track it, but right now it's just not something we're focusing on with 1.4 around the corner. It's like trying to go buy a house in Florida when there's a hurricane coming on shore: we'd rather prepare for the hurricane.

**Comment by hakusaro** [comment]
Also, and this is important, only our developers from 🇨🇳 have mobile versions laying around. @mistzzt or @AxeelAnder might be better inclined to comment?

I personally have bought like 9 copies of Terraria already and I'd really like to not have to buy another copy.

**Comment by hakusaro** [comment]
In response to this issue, I have bought a copy of Terraria for mobile.

**Reply by tylerjwatson** [reply]
It seems like maintaining a 1.3.0.7 fork is counterintuitive giving that the patches for the latest version are already handled by us and Terraria mobile is likely to just update at any point?

**Reply by hakusaro** [reply]
Yes, I agree. I don't intend to do anything other than some mobile reconnaissance. If 1.4 drops for mobile and we have version parity it's not worth the time/investment to try and go back to 1.3.0.7.

**Reply by tylerjwatson** [reply]
If you think it's worth looking into please let me know and I will buy a copy.

**Reply by bartico6** [reply]
It was confirmed that mobile will indeed get 1.4 this year and other versions before that probably. I don't know how long it'll take but from my POV, Mobile 1.3.0.7 has been incredibly painful to work with because of how its built.

**Reply by moisterrific** [reply]
> Mobile 1.3.0.7 has been incredibly painful to work with because of how its built.

This tbh, I worked on maintaining the mobile fork originally by Axeel and it was a huge headache trying to fix the issues. Although I’m not a coder by trade but even I could see it gets real messy trying to make the latest 4.3.26 TShock run with older 1.3.0.7.

Hopefully when 1.4 hits mobile it will be as easy as handling the mobile MaxSlots difference when adding cross-platform compatibility. But in the meantime it’s not really worthwhile, majority of players are on PC anyways. 

@Nekit0zz there is a plugin called Dimensions that can enable backwards compatibility with mobile on the latest PC TShock, it’s what Terraria.one and a few other uses. Maybe you can give that a try, still quite buggy for mobile from my experience tho

**Comment by Brx86** [comment]
Actually i have run a mobile tshock server for half a year. As far as I know, there are several custom editions of tshock mobile work well in China. I'm sorry that I can't provide it for you, but maybe you can have a try at this: https://github.com/Fe7n/TShockMobile

**Reply by hakusaro** [reply]
@Brx86 are you running a custom port you made or were you speaking entirely past tense?

**Reply by AxeelAnder** [reply]
> @Brx86 are you running a custom port you made or were you speaking entirely past tense?

@qq1176321897 manually patched my mobile fork, he fixed several bugs exist in my fork. His version is widely used in china.

**Reply by Brx86** [reply]
> @Brx86 are you running a custom port you made or were you speaking entirely past tense?

It is based on tshock43260, but not made by myself.

**Comment by hakusaro** [comment]
In an attempt to clear up as much confusion as possible, I've decided to make a FAQ.

## Is there an officially supported version of TShock for Terraria on mobile?

No.

### What do you mean by officially supported?

I mean, like, we, @Pryaxis, don't support or give away copies in the sense that we don't feel obligated to fix issues with it. We love people who work on these ports! People who work on stuff for other people are really sweet and we really want to see more of this happening.

### Why is there no version of TShock for mobile Terraria yet?

Honestly, because we haven't been really interested in building it yet. That may change though, so don't worry too much.

## Is it possible for there to be a mobile version of TShock?

Yes. Lots of people have done it, to varying levels of success. Some people are more successful than others. Nothing is impossible. An off the shelf copy of TShock for that version of Terraria is unlikely to be successful, though, because...

## What stops anyone from just running TShock for Terraria on mobile?

Mobile versions of Terraria are on an older version of Terraria than the current PC release.

Back when Zidonuke was around, he kept a record of the Terraria multiplayer protocol exploits in a huge spreadsheet. It was a huge list of every single multiplayer bug, exploit, and mitigation that TShock had in place. Over time, we've been able to send information to Re-Logic to patch the bugs in the official game, or patch them ourselves. This is why, for example, the Drill Containment Unit isn't completely broken and totally stopped in TShock anymore -- it got patched at the protocol level.

Because all of the mobile versions are old, they contain all of the original exploits that we tried to get fixed as well as new ones specific to mobile. As a result, you have a very high risk that something bad happens if you run a server on mobile without going through the effort of doing security updates and testing specifically on mobile.

## What about 1.4 / Journey's End?

It's been rumoured that when Journey's End / 1.4 comes out, mobile will be updated soon after so that it's on the same game version. If/when this happens, it'll be easy for TShock to support mobile officially. Will it happen for sure? I can't really make any promises. Is it possible? Definitely. Have we heard interest? Oh yeah.
